infer/lib/python/inferlib/capture/gradle.py

```diff
@@ -31,9 +29,92 @@ def gen_instance(*args):
 create_argparser = util.base_argparser(MODULE_DESCRIPTION, MODULE_NAME)
 
 
+def extract_filepath(parts):
+    if not parts:
+        return ([], None)
+    path = ' '.join(parts)
+    if os.path.isfile(path):
+        return ([], path)
+    remainder, path = extract_filepath(parts[1:])
+    return ([parts[0]] + remainder, path)
+
+
+# Please run the doctests using:
+# $ python -m doctest -v gradle.py
+def extract_all(javac_arguments):
+    """Extract Java filenames and Javac options from the Javac arguments.
+
+    >>> os.path.isfile = lambda s: s[1:].startswith('path/to/')
+    >>> extract_all([])
+    {'files': [], 'opts': []}
+    >>> extract_all(['-opt1', 'optval1', '/path/to/1.java'])
+    {'files': ['/path/to/1.java'], 'opts': ['-opt1', 'optval1']}
+    >>> extract_all(['-opt1', 'optval1', '/path/to/a', 'b/1.java'])
+    {'files': ['/path/to/a b/1.java'], 'opts': ['-opt1', 'optval1']}
+    >>> extract_all(['-opt1', 'opt', 'val1', '/path/to/1.java'])
+    {'files': ['/path/to/1.java'], 'opts': ['-opt1', 'opt val1']}
+    >>> extract_all(['-opt1', '/path/to/a', 'b/c', 'd/1.java', '-opt2'])
+    {'files': ['/path/to/a b/c d/1.java'], 'opts': ['-opt1', '-opt2']}
+    >>> extract_all(['-opt1', 'optval1', '-path/to/1.java'])
+    {'files': ['-path/to/1.java'], 'opts': ['-opt1', 'optval1']}
+    >>> extract_all(['-opt1', 'optval1', '/path/to/', '-1.java'])
+    {'files': ['/path/to/ -1.java'], 'opts': ['-opt1', 'optval1']}
+    >>> extract_all(['undef1', 'undef2'])
+    {'files': [], 'opts': ['undef1 undef2']}
+    """
+    def pop():
+        if javac_arguments:
+            return javac_arguments.pop()
+        return None
+
+    java_files = []
+    java_opts = []
+    # Reversed Javac options parameters
+    rev_opt_params = []
+    java_arg = pop()
+    while java_arg:
+        if java_arg.endswith('.java'):
+            # Probably got a file
+            remainder, path = extract_filepath(javac_arguments + [java_arg])
+            if path:
+                java_files.append(path)
+                javac_arguments = remainder
+            else:
+                # A use-case here: *.java dir as an option parameter
+                rev_opt_params.append(java_arg)
+        elif java_arg[0] == '-':
```

Comment:
This cannot happen because Gradle doesn't escape anything it gives us, but better use `java_arg.startswith('-')` in case `java_arg` has `0` length, to avoid crashing.